lib/instrumentation/modules/memcached.js

```diff
@@ -0,0 +1,49 @@
+'use strict'
+
+var semver = require('semver')
+
+var shimmer = require('../shimmer')
+
+module.exports = function (memcached, agent, { version, enabled }) {
+  if (!enabled) {
+    return memcached
+  }
+  if (!semver.satisfies(version, '>=2.2.0')) {
+    agent.logger.debug('Memcached version %s not supported - aborting...', version)
+    return memcached
+  }
+  agent.logger.debug('shimming memcached.prototype.command')
+  shimmer.wrap(memcached.prototype, 'command', wrapCommand)
+  return memcached
+
+  // Wrap the generic command that is used to build touch, get, gets etc
+  function wrapCommand (original) {
+    return function wrappedCommand () {
+      if (typeof arguments[0] === 'function') {
+        var query = arguments[0]()
+        // If the callback is not a function the user doesn't care about result
+        if (query && typeof query.callback === 'function') {
+          var span = agent.startSpan(`memcached.${query.type}`, `db.memcached.${query.type}`)
+          if (span) {
+            span.setDbContext({ statement: `${query.type} ${query.key}`, type: 'memcached' })
+            var id = span.id
+          }
+          agent.logger.debug('intercepted call to memcached.prototype.command %o', { id: id, type: query.type })
+          query.callback = wrapCallback(query.callback)
+          // Rewrite the query compiler with the wrapped callback
+          arguments[0] = function queryCompiler () {
```

Comment:
Does this need to be outside of the `if (query && typeof query.callback === 'function')`? If you're trying to avoid side effects from calling the function multiple times, it won't work if that check fails. I'm not really sure what the purpose of this wrap here is though. 🤔

Reply:
It does not need to be outside. We only need to change arguments[0] when we wrap the callback. I'm not worried about it being called multiple times because it doesn't mutate state. Their argument[0] returns a static object with a variable callback(set by user)

Reply:
I think it makes sense to not call the query compiler more than once, so I like the pattern. But yeah, it should probably be outside of the if block - good catch! I'd probably put it right after the line where the `query` variable is set so that the two are next to each other. That makes it easier to understand IMO.

Reply:
@mikefaut Ups, sorry. Didn't see your comment when I wrote mine. GitHub sometimes doesn't update very well 😅

Reply:
Now reading your comment, I agree that it's fine to leave inside for the current implementation of `memcached`. And though they probably will never change this behavior, wouldn't it be more future proof / safe to put it outside of the if-block? I'm fine either way though.